core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageDrainer.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+import io.netty.util.ReferenceCountUtil;
+
+final class StreamMessageDrainer<T> implements Subscriber<T> {
```

Comment:
You could make this class non-final and add some extension point so that a subclass can add an element selectively for `FilteredStreamMessage`, e.g.

```java
class StreamMessageDrainer implements Subscriber {
    ...
    final add(T t) {
        drained.add(t);
    }
}

class FilteredStreamMessageDrainer extends StreamMessageDrainer {
    @Override
    public void onNext(T t) {
        if (...) {
            add(t);
        } else {
            ReferenceCountUtil.safeRelease(t);
        }
    }
}
```

Reply:
I realized that we don't need this because we're already using `FilteringSubscriber`. My initial approach was just wrong. Thank you for letting me wake up!